pandas/core/groupby.py

```diff
@@ -4585,9 +4645,17 @@ def _apply_to_column_groupbys(self, func):
              in self._iterate_column_groupbys()),
             keys=self._selected_obj.columns, axis=1)
 
+    def _fill(self, direction, limit=None):
+        """Overriden method to join grouped columns in output"""
+        res = super(DataFrameGroupBy, self)._fill(direction, limit=limit)
+        output = collections.OrderedDict(
+            (grp.name, grp.group_index.take(grp.labels)) for grp in
```

Comment:
maybe missing something obvious, but isn't the `take` unnecessary - couldn't you use the original grouping columns?

Reply:
Certainly could be a better way to do it - first time working this far into the API so if there's any easier way you know of that I missed let me know.

FWIW I think that instead of the `take` we could replace this with `grp.grouper` and get the same values - is that what you are referring to?

Reply:
Sorry for the delay - was working from my memory and I honestly get lost on all the levels of groupby, groupers, and groupings.  What you want here is `grp.grouper` - it has the original values without the need to take again.

```python
In [40]: df = pd.DataFrame({'k': [1, 1, 2, 2, 1, 1], 'v': [1., np.nan, 2., np.nan, np.nan, 3.]})

In [41]: gb = df.groupby('k')

In [42]: [grp.grouper for grp in gb.grouper.groupings]
Out[42]: [array([1, 1, 2, 2, 1, 1], dtype=int64)]
```